Allow hostname passing

RRRadicalEdward opened the pull request:
I've improved `sspi-rs` API and now we can pass the hostname directly into the `SspiContext`.

I added a new trait method `SspiEx::custom_set_hostname` and implemented it for all the security packages. I decided to not change the main `Sspi` trait because the original `SSPI` doesn't have such a method. On non-wasm targets, we still use `whoami` crate because it's needed for all other applications (like `FreeRDP`) that work with `sspi-rs` using FFI.

CBenoit commented on Cargo.toml:
```diff
@@ -55,6 +55,9 @@ portpicker = { version = "0.1.1", optional = true }
 num-bigint-dig = "0.8.1"
 tracing = { version = "0.1.37" }
 
+[target.'cfg(not(target_arch = "wasm32-unknown-unknown"))'.dependencies]
+whoami = "0.5"
+
```
Ideally, this dependency should be in `sspi-ffi` only and `custom_set_hostname` should be called with the result.

RRRadicalEdward replied:
Done

CBenoit reviewed (commented):
I was thinking the hostname would be a parameter when building  `CredSspClient` and the `whoami` dependency part of the `sspi-ffi` crate. Do you think that would be acceptable?

RRRadicalEdward commented:
Hi @CBenoit! I added the option.

The hostname is only required for KDC/KerberosProxy. Also, I noticed that you wouldn't be able to negotiate a protocol in WASM. As KDC/KerberosProxy was behind the `network_client` feature that is not accessible on WASM(hardcoded usage of the `reqwest` client) in the `negotiate_protocol` function. So, I refactored the `negotiate_protocol` function to make it work on WASM. I created a separate `RequestClientFactory` trait, so you can pass your client in WASM. In FFI, it still uses the `reqwest` client.

CBenoit commented on tools/wasm-testcompile/src/lib.rs:
```diff
@@ -1,14 +1,33 @@
 use wasm_bindgen::prelude::*;
 
+#[derive(Debug, Clone)]
+struct DumpNetworkClientFactory;
```
I’m unsure of the name. Do you mean "Dummy" here?

```suggestion
struct DummyNetworkClientFactory;
```

RRRadicalEdward replied:
Yeah 😄

RRRadicalEdward replied:
Fixed

CBenoit commented on src/sspi/negotiate/network_client.rs:
```diff
@@ -1,7 +1,14 @@
+use std::fmt::Debug;
```
I see you moved this module under `negotiate`. I understand this can be used at a few places (including kerberos) and it’s generic enough so I feel it could be at top level, what do you think?

CBenoit commented on src/sspi/negotiate.rs:
```diff
@@ -43,29 +41,35 @@ pub trait ProtocolConfig: Debug {
             protocol_config,
             package_list,
+            hostname,
+            network_client_generator: get_network_client,
```
I feel it would improve clarity to keep the same terminology at both places.
```suggestion
            network_client_factory: get_network_client,
```

CBenoit reviewed (commented):
Nice improvements, thank you!

CBenoit reviewed (approved):
LGTM!
